### PR TITLE
added `heroku dashboard` and related commands

### PR DESCRIPTION
**Pull request by dickeyxxx**
This is experimental (but I encourage you to try it if you want!)

screen shot 2016-04-06 at 2 02 21 pm

The goal is not to add a new command `heroku dashboard` (it will probably be there but hidden), but to change how `heroku` works with no arguments to run this command.

TODO:

- [x] test `dashboard`
- [x] test `notifications`
- [x] test `apps:errors`
- [x] test `apps:favorites`
- [x] test `apps:favorites:add`
- [x] test `apps:favorites:remove`

### CODE REVIEW

**Comment by raulb** [comment]
I love the fact that we're leveraging favorites more!

**Comment by raulb** [comment]
Any plans on making favorites part of the public platform API?

File: commands/apps/dashboard.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  if (metrics.routerStatus && !empty(metrics.routerStatus.data)) {
+    rpm = metrics.routerStatus  ? `${_.round(_.sum(_.flatten(_.values(metrics.routerStatus.data)))/24/60)} rpm` : '';
+  }
+  cli.log(`    ${cli.color.green(ms)} ${cli.color.yellow(rpm)}`);
+}
+
+function* run (context, heroku) {
+  function favoriteApps () {
+    return heroku.request({
+      host: 'longboard.heroku.com',
```
Comment (paulelliott) [inline commands/apps/dashboard.js]: Longboard is not a public API and is subject to change without notice.

Comment (dickeyxxx) [inline commands/apps/dashboard.js]: I'd really like to use it for this, any suggestions on what you think the right way to go is?

Comment (paulelliott) [inline commands/apps/dashboard.js]: Unfortunately this is the only way to get at that data today. I don't think this is a bad idea, we just need to be aware that you're consuming it and make you aware of any changes.

We want to do some refactoring of it but that work is not scheduled and likely won't happen in the near term. I would recommend watching the https://github.com/heroku/uku-favorites project for updates.

Comment (dickeyxxx) [inline commands/apps/dashboard.js]: sounds good, @paulelliott I might see if I can schedule a meeting to chat with you a bit more about this. Essentially what my goal here is is to bring in some of the missing dashboard functionality into the CLI.

**Comment by dickeyxxx** [comment]
@raulb I'm not aware of it, but that would definitely be nice!

**Comment by halorgium** [comment]
THANKS!!!! ❤️ 💛 ❤️

File: commands/dashboard/index.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    notifications: heroku.request({host: 'telex.heroku.com', path: '/user/notifications'}),
+    apps:    apps.map(app => ({
+      app:         heroku.get(`/apps/${app}`),
+      formation:   heroku.get(`/apps/${app}/formation`),
+      pipeline:    heroku.get(`/apps/${app}/pipeline-couplings`).catch(() => null),
+    })),
+  };
+
+  if (apps.length > 0) yield co(displayApps(data.apps));
+  else cli.warn(`Add apps to this dashboard by favoriting them with ${cli.color.blue('heroku favorites:add')}`);
```
Comment (ransombriggs) [inline commands/dashboard/index.js]: @dickeyxxx This command does not exist, I assume you want `apps:favorites:add`

File: commands/apps/errors.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    dyno:    _.fromPairs(types.map(type => [type, dynoErrors(type)])),
+    router:  routerErrors(),
+  };
+
+  if (context.flags.json) {
+    cli.styledJSON(errors);
+  } else {
+    let t = buildErrorTable(errors.router, 'router');
+    for (let type of Object.keys(errors.dyno)) t = t.concat(buildErrorTable(errors.dyno[type]));
+    cli.table(t, {
```
Comment (ransombriggs) [inline commands/apps/errors.js]: @dickeyxxx When there are no errors, this bombs out with
```
 ▸    Cannot read property 'length' of undefined
TypeError: Cannot read property 'length' of undefined
```

Comment (dickeyxxx) [inline commands/apps/errors.js]: fixed

File: commands/dashboard/index.js (hunk 2 above)
Comment (dickeyxxx) [inline commands/dashboard/index.js]: fixed

File: commands/dashboard/index.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      pipeline:    heroku.get(`/apps/${app}/pipeline-couplings`).catch(() => null),
+    })),
+  };
+
+  if (apps.length > 0) yield co(displayApps(data.apps));
+  else cli.warn(`Add apps to this dashboard by favoriting them with ${cli.color.blue('heroku apps:favorites:add')}`);
+
+  cli.log(`\nSee all add-ons with ${cli.color.blue('heroku addons')}`);
+  let sampleOrg = _.sortBy(data.orgs, o => new Date(o.created_at))[0];
+  if (sampleOrg) cli.log(`See all apps in ${cli.color.blue(sampleOrg.name)} with ${cli.color.blue('heroku apps --org ' + sampleOrg.name)}`);
```
Comment (ransombriggs) [inline commands/dashboard/index.js]: @dickeyxxx I was testing this and it listed an org that I am a collaborator on, but do not have access to the app listing to.  We should narrow this down to just those that would actually allow app listings.  Holler at me in chat and I can find the code in the backend that matches.

Comment (dickeyxxx) [inline commands/dashboard/index.js]: put a patch in for that, let me know if you have a better solution though

**Comment by ransombriggs** [comment]
@dickeyxxx lgtm

**Pull request merged**